src/dispatch/stage.rs

```diff
@@ -0,0 +1,438 @@
+
+        let (stage, group) = match target {
+            InsertionTarget::Stage(stage) => {
+                let group = self.ids[stage].len();
+                self.add_group(stage);
+
+                (stage, group)
+            }
+            InsertionTarget::Group(stage, group) => (stage, group),
+            InsertionTarget::NewStage => {
+                let stage = self.stages.len();
+
+                self.add_stage();
+                self.add_group(stage);
+
+                (stage, 0)
+            }
+        };
+
+        self.ids[stage][group].push(id);
+        self.reads[stage][group].extend(reads);
+        self.running_time[stage][group] += new_time as u8;
+        self.stages[stage].groups[group].push(Box::new(system));
+        self.writes[stage][group].extend(writes);
+    }
+
+    pub fn build(self) -> Vec<Stage<'a>> {
+        self.stages
+    }
+
+    fn add_stage(&mut self) {
+        self.ids.push(GroupVec::new());
+        self.reads.push(GroupVec::new());
+        self.running_time.push(GroupVec::new());
+        self.stages.push(Stage::new());
+        self.writes.push(GroupVec::new());
+    }
+
+    fn add_group(&mut self, stage: usize) {
+        self.ids[stage].push(ArrayVec::new());
+        self.reads[stage].push(SmallVec::new());
+        self.running_time[stage].push(0);
+        self.stages[stage].groups.push(ArrayVec::new());
+        self.writes[stage].push(SmallVec::new());
+    }
+
+    fn insertion_target<'rw, R, W>(&self,
+                                   new_reads: R,
+                                   new_writes: W,
+                                   new_dep: &mut SmallVec<[SystemId; 4]>,
+                                   new_time: RunningTime)
+                                   -> InsertionTarget
+        where R: IntoIterator<Item = &'rw ResourceId>,
+              R::IntoIter: Clone,
+              W: IntoIterator<Item = &'rw ResourceId>,
+              W::IntoIter: Clone
+    {
+        let new_reads = new_reads.into_iter();
+        let new_writes = new_writes.into_iter();
+
+        (self.barrier..self.stages.len())
+            .map(|stage| {
+                let conflict = Self::find_conflict(&*self.ids,
+                                                   &*self.reads,
+                                                   &*self.writes,
+                                                   stage,
+                                                   new_reads.clone(),
+                                                   new_writes.clone(),
+                                                   new_dep);
+                self.remove_ids(stage, new_dep);
+                (stage, conflict)
+            })
+            .find(|&(stage, conflict)| match conflict {
+                      Conflict::None => true,
+                      Conflict::Single(group) => {
+                          self.stages[stage].groups[group].len() < MAX_SYSTEMS_PER_GROUP - 1 &&
```

Comment:
The `ArrayVec` only supports up to `MAX_SYSTEMS_PER_GROUP`, so any additional system would just be discarded. So I check if there's enough space left.